Tests/test_file_gif.py

```diff
@@ -507,6 +505,35 @@ def test_save_I(self):
         reloaded = Image.open(out)
         self.assert_image_equal(reloaded.convert('L'), im.convert('L'))
 
+    def test_getdata(self):
+        # test getheader/getdata against legacy values
+        # Create a 'P' image with holes in the palette
+        im = Image._wedge().resize((16,16))
+        im.putpalette(ImagePalette.ImagePalette('RGB'))
+        im.info = {'background':0}
+        
+        passed_palette = bytes(bytearray([255-i//3 for i in range(768)]))
+
+        GifImagePlugin._FORCE_OPTIMIZE = True
+        try:
+            h = GifImagePlugin.getheader(im, passed_palette)
+            d = GifImagePlugin.getdata(im)
+
+            import cPickle as pickle
+            # Enable to get target values on pre-refactor version
+            # with open('Tests/images/gif_header_data.pkl', 'wb') as f:
+            #    f.write(pickle.dumps((h,d)))
+            with open('Tests/images/gif_header_data.pkl', 'rb') as f:
+                (h_target,d_target) = pickle.loads(f.read())
```

Comment:
So, ignoring pep8 for a moment, this is a big PR and needs real review. 

This fixes at least a couple longstanding bugs, splits out image modification into explicit steps and retains compatibility with the earlier version of the legacy images.  Now that getheader/getdata are tests, I think we've got everything covered so we don't break anything. There is still at least one corner case where I'm not sure _why_ we're doing what we're doing, (the initial palette thing for 'RGB' base mode images). 

- Is this pythonic?  ey, maybe. could be better. 
- Is it clearer than it was?
- Is there anything that's confusing in the code?